lib/elixir/lib/map_set.ex

```diff
@@ -84,13 +94,49 @@ defmodule MapSet do
 
   """
   @spec difference(t, t) :: t
+  # If the first set is less than twice the size of the second map,
+  # it is fastest to re-accumulate items in the first set that are not
+  # present in the second set.
+  def difference(%MapSet{map: map1}, %MapSet{map: map2})
+    when map_size(map1) < map_size(map2) * 2 do
+
+    map = map1
+    |> Map.keys
+    |> filter_not_in(map2)
+
+    %MapSet{map: map}
+  end
+
+  # If the second set is less than half the size of the first set, it's fastest
+  # to simply iterate through each item in the second set, deleting them from
+  # the first set.
   def difference(%MapSet{map: map1}, %MapSet{map: map2}) do
-    map = :maps.fold(fn value, _, acc ->
-      Map.delete(acc, value)
-    end, map1, map2)
+    map = map2
+    |> Map.keys
+    |> delete_from(map1)
+
     %MapSet{map: map}
   end
 
+  def delete_from([], acc), do: acc
+  def delete_from([k | rest], acc) do
+    delete_from(rest, Map.delete(acc, k))
+  end
```

Comment:
I should honestly re-write drop to use manual recursion now that I've discovered it gives us ~20% improvement in cases like this. Then yes we could just re-write this as `Map.drop(map1, Map.keys(map2))`